[로또] 러너덕 미션 제출합니다.

Deocksoo opened the pull request:
제출이 너무 늦어 죄송합니다 ㅠㅠ 중간에 이상한 방향으로 고민을 조금 오래 하다가 시간이 너무 지체되었네요. 피드백 주시면 빠르게 반영하도록 하겠습니다!

Deocksoo committed:
[DOCS] 로또 기능 목록을 추가합니다

Deocksoo committed:
[ADD] LottoNumberTest와 구현사항을 추가합니다

Deocksoo committed:
[ADD] LottoTest와 구현사항을 추가합니다

Deocksoo committed:
[REFACTOR] LottoTest field 의 접근제어자를 변경합니다

Deocksoo committed:
[ADD] 로또에 숫자 매칭 기능을 추가합니다

winning lotto와 숫자를 맞춰보고 일치하는 숫자에 따라 매칭 결과를 되돌려주도록
구현하였습니다. 매칭 결과는 3~6의 숫자 중 하나이므로, enum을 통해 상수로
구현하였습니다.

Deocksoo committed:
[REFACTOR] Lotto 일급 객체의 자료 구조를 변경합니다

List 대신 Set을 사용해 중복되는 숫자가 들어가지 못하게 변경하였습니다.

Deocksoo committed:
[ADD] 구입 금액 객체를 추가합니다

Deocksoo committed:
[ADD] 당첨 통계 객체를 추가합니다

당첨 결과별 당첨 통계를 셈하기 위해 CountOfRank 클래스를 추가하였습니다.

Deocksoo committed:
[ADD] MatchingStatistic 클래스에 총 수익률을 계산하는 기능을 추가합니다

Deocksoo committed:
[REFACTOR] MatchingStatistics 클래스명을 Statistics로 변경합니다.

더불어 Statistics클래스의 calculateEarningRatesOf 메서드를 리펙토링하였습니다

Deocksoo committed:
[REFACTOR] Statistics 클래스의 add와 contsOf 메서드의 argument명을 변경합니다

Deocksoo committed:
[REFACTOR] Purchasement 클래스명을 PurchaseAmount로 변경합니다

구입 금액의 의미를 명시하기 위해 이름을 변경하였습니다.

Deocksoo committed:
[EDIT] Rank 클래스 변수의 오타 수정

Deocksoo committed:
[ADD] 구입 금액을 입력받는 기능을 추가합니다

Deocksoo committed:
[REFACTOR] Lotto의 가격 정보를 Lotto Class field로 이동합니다

Deocksoo committed:
[ADD] LottoFactory를 추가합니다

LottoNumber instance를 중복하여 생성하지 않게 하기 위하여 LottoFactor의
field에 static 상수로 할당한 뒤 사용하도록 하였습니다.

Deocksoo committed:
[ADD] 발행된 lotto를 출력해주는 기능을 추가합니다

Deocksoo committed:
[ADD] LottoFactory에 당첨 로또를 만드는 기능을 추가합니다

Deocksoo committed:
[STYLE] PurchseAmount 클래스의 메소드 순서를 변경합니다

가독성을 위해 생성자에서 호출하는 순서에 맞게 수정하였습니다.

Deocksoo committed:
[ADD] LottoGameLauncher에 winningLotto 생성 로직을 추가합니다

Deocksoo committed:
[ADD] 당첨 통계와 총 수익률을 출력하는 기능을 추가합니다

실제 Lotto 추첨을 LottoGame이 담당하도록 하고, 추첨 결과 통계를 리턴하도록
구성하였습니다.

Deocksoo committed:
[ADD] 당첨 번호에 보너스 번호를 추가합니다

보너스 번호가 추가된 당첨 로또를 구분하기 위하여 WinningLotto와 issuedLotto로 나누어
구현하였습니다.

Deocksoo committed:
[REFACTOR] LottoNumberPool을 LottoFactory로부터 분리합니다

Deocksoo committed:
[REFACTOR] LottoNumber와 관련된 클래스들을 패키지로 묶습니다

Deocksoo committed:
[ADD] LottoFactory에 수동 발행 기능을 추가합니다

Deocksoo committed:
[ADD] IssuedLottos 일급 콜렉션을 추가합니다

Deocksoo committed:
[ADD] 수동 구매 기능을 추가합니다

Deocksoo committed:
[REFACTOR] InputView를 리펙토링합니다.

Deocksoo committed:
[REFACTOR] 불필요한 접근 제어자와 space를 수정합니다

Deocksoo committed:
[REFACTOR] 2등은 OutputView에서 다르게 출력되도록 수정합니다

Deocksoo committed:
[FIX] IssuedLottos 일급콜렉션의 getTotalLottosOf 메서드 오타 수정

두 lotto 그룹을 합치는 과정에서 auto를 manual로 잘못 표기하여 이를 수정합니다.

Deocksoo committed:
[REFACTOR] LottoGameLaucher클래스의 getPurchaseAmount메소드 위치를 변경합니다

woowahan-pjs commented on src/main/java/LottoGameLauncher.java:
```diff
@@ -0,0 +1,76 @@
+        try {
+            int moneyFromPlayer = InputView.inputPurchaseAmount();
+            return PurchaseAmount.valueOf(moneyFromPlayer);
+        } catch (IllegalArgumentException e) {
```
캡틴께서 "반복문 대신 재귀 함수로 구현할 수 있다."라고 알려주셨어요. 하지만 재귀를 사용하면 어떤 위험이 있을까요? 만약 재귀를 사용하지 않는다면 어떻게 문제를 해결할 수 있을까요?

woowahan-pjs commented on README.md:
```diff
@@ -1,2 +1,30 @@
+     - 일치하는 숫자별로 몇 개의 로또가 당첨되었는지 알 수 있어야 합니다.
+     - 일치하는 숫자별 당첨 금액을 알 수 있어야 합니다.
+     - 총 수익률을 알 수 있어야 합니다.
+   * 수동 구매 기능을 추가해야 합니다.
```
꼼꼼한 README.md 파일 💯

woowahan-pjs commented on src/main/java/domain/IllegalNumberOfManualIssueException.java:
```diff
@@ -0,0 +1,7 @@
+package domain;
+
+public class IllegalNumberOfManualIssueException extends IllegalArgumentException {
```
커스터마이징 한 예외 클래스 👍

woowahan-pjs commented on src/main/java/domain/CountOfRank.java:
```diff
@@ -0,0 +1,15 @@
+package domain;
+
+class CountOfRank {
```
생성자를 통해 접근을 제어하는 것과 어떤 차이가 있을까요? 같이 생각해봐요.

woowahan-pjs commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Arrays;
+
+public enum Rank {
+    FIRST(6, 2000000000),
```
Java7부터는 `100_000_000`처럼 숫자에 밑줄을 사용하여 가독성을 높일 수 있어요.

woowahan-pjs commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,50 @@
+    SECOND(5, 30000000) {
+        @Override
+        public String getRankExplanation() {
+            return SECOND.numberOfMatching + "개 일치, 보너스 볼 일치 "
```
Model에 View의 요구 사항을 반영하는 것이 맞을까요? Model과 View의 기능을 분리하면 어떤 장점이 있을까요?

woowahan-pjs commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,50 @@
+
+public enum Rank {
+    FIRST(6, 2000000000),
+    SECOND(5, 30000000) {
```
"2등은 보너스 번호가 있다."라는 요구 사항을 코드로 표현해 보면 어떨까요?

woowahan-pjs commented on src/main/java/domain/lottonumber/LottoNumber.java:
```diff
@@ -0,0 +1,52 @@
+        return Integer.toString(number);
+    }
+
+    boolean isOf(int number) {
```
`equals()`, `hashCode()`, `toString()` 메서드가 마지막에 위치하는 것이 일반적인 관례입니다.

woowahan-pjs commented on src/main/java/domain/Statistics.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    static Statistics of(Rank[] values) {
+        EnumMap<Rank, CountOfRank> countsOfRanks = new EnumMap<>(Rank.class);
```
인터페이스와 다형성에 대해 알아보고 아래의 글도 참고하면 좋을 것 같아요.
https://stackoverflow.com/questions/11683044/return-arraylist-or-list/11683073

woowahan-pjs commented on src/main/java/domain/Statistics.java:
```diff
@@ -0,0 +1,50 @@
+            winningMoneyOfRank = rank.getWinningMoney();
+            totalWinningMoney += countsOf(rank) * winningMoneyOfRank;
+        }
+        return totalWinningMoney / (double) purchasedAmount;
```
아래의 글은 가볍게 참고만 해주세요.
https://isooo.github.io/java/2019/01/17/BigDecimal.html

woowahan-pjs reviewed (changes requested):
오프라인 코드 리뷰로 대신하였습니다. 👍

Deocksoo committed:
[REFACTOR] CountOfRank 접근제어자를 변경합니다

Deocksoo committed:
[REFACTOR] InputView 오타를 수정합니다

Deocksoo committed:
[REFACTOR] Rank의 출력메시지를 output view에서 담당하도록 변경합니다

Deocksoo committed:
[REFACTOR] Statistics 클래스에서 add 메서드를 분리합니다

LottoGame 클래스에서 미리 EnumMap을 만들고, Rank별 개수를 count한 뒤
Statistics 객체를 만들어 리턴해주도록 변경하였습니다.

Deocksoo committed:
[STYLE] LottoGameLauncher 클래스 코드 오토 포매팅

Deocksoo committed:
[REFACTOR] InputView에서 parsing 기능을 분리합니다.

StringParser 클래스를 만들어 분리하였습니다.

Deocksoo committed:
[REFACTOR] InputView에서 분리된 parsing기능을 LottoGameLauncher에 반영합니다

Deocksoo committed:
[REFACTOR] PurchaseAmount와 Money를 분리합니다

PurchaseAount로 모든 금액을 포장할 경우 다음과 같은 문제들이 있습니다.
1. purhcaseAmount객체에서 1000원 이상만 생성되도록 하므로, 수동 구매한
로또 개수가 0일 때 구매 금액을 다시 역산하면 예외가 던져집니다.
2. 도메인 내의 다른 돈(ex. Rank의 WinningMoney)을 포장할 객체가 또 필요합니다.

위와 같은 이유로 Money 객체를 분리하였습니다.

Deocksoo committed:
[REFACTOR] LottoNumberPool의 캐시를 LottoNumber로 이동합니다

1. Java Integer 클래스의 캐시 위치를 보고 판단했을 때 LottoNumber에서 캐싱의 역할을
맡는 것이 더 낫다고 판단해 이동하였습니다.
2. 기존에 LottoNumberPool에서 로또 숫자를 랜덤하게 뽑아주던 기능을 고려하여
캐시를 뽑아내고 남은 기능은 RandomLottoNumberGenerator로 리펙토링 하였습니다.

Deocksoo committed:
[REFACTOR] Money와 관련된 클래스들을 money 패키지로 만듭니다

Deocksoo committed:
[REFACTOR] Rank의 winningMoney를 Money객체로 포장합니다

1. int -> Money로 데이터 타입을 변경하였습니다.
2. Money 클래스의 valueOf 펙토리 메서드 이름을 amountOf로 변경하였습니다.

Deocksoo committed:
[REFACTOR] Rank의 펙토리 메서드를 수정합니다

Rank의 instance들을 생성할 때, bonus 숫자의 매칭 여부를 생성시 명시적으로 드러날
수 있게 하기 위해 전략 패턴을 사용하였습니다.

Deocksoo committed:
[REFACTOR] WinningLotto의 MatchUpLottoNumbersWith메서드를 리펙토링합니다

Deocksoo committed:
[REFACTOR] WinningLotto와 Lotto의 연관관계를 변경합니다

상속 관계에서 포함 관계로 변경합니다.

Deocksoo committed:
[REFACTOR] LottoFactory의 getLottoNumberOf 메서드를 LottoNumber로 이동합니다

LottoNumber와 관련된 기능이라 수정하였습니다.

Deocksoo committed:
[FIX] 새로 만든 Lotto 클래스에 toString 메서드를 추가합니다

Deocksoo committed:
[REFACTOR] Lotto PRICE를 Money 객체로 포장합니다

원시값이었던 Lotto PRICE를 Money 객체로 포장하였습니다. 그러는 과정에서
Money 객체에 multiply 메서드가 새로 필요하게 되어 추가하였습니다.

Deocksoo committed:
[REFACTOR] Statistics 객체가 생성될 때 enumap을 새로 만들어 생성하도록 수정하였습니다

Deocksoo committed:
[REFACTOR] RandomLottoNumberGenerator에 private 생성자를 추가합니다

singleton 객체로서 static 함수만 사용하게 하기 위해 private 생성자를 추가하였습니다.

Deocksoo committed:
[REFACTOR] earningRates를 Controller에서 계산하도록 수정합니다

woowahan-pjs reviewed (approved):
피드백을 모두 잘 반영해주셨습니다. 💯 
다음 단계를 진행하려면 woowacouse에 Merge 된 코드에서 다시 시작해야 충돌이 발생하지 않을 거예요.
아래의 과정을 진행한 후, 다음 단계 진행할 것을 추천해요.

```
1. git remote add upstream
2. git fetch upstream 자신의 github_id
3. git reset --hard upstream/자신의 github_id
```

다음 단계에서 뵙겠습니다. 이만 Merge 하겠습니다.
